lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs

```diff
@@ -0,0 +1,169 @@
+
+-- | Encrypt using AES256 using CBC mode
+encrypt
+    :: CipherMode
+    -- ^ with or without padding
+    -> ByteString
+    -- ^ secret key, needs to be 32 bytes
+    -> ByteString
+    -- ^ iv, needs to be 16 bytes
+    -> ByteString
+    -- ^ payload, must be a multiple of a block size, ie., 16 bytes
+    -> Either CipherError ByteString
+encrypt mode key iv msg = do
+   when (mode == WithoutPadding && BS.length msg `mod` 16 /= 0) $
+       Left $ WrongPayloadSize
+   initedIV <- mapLeft FromCryptonite (initIV iv)
+   let msgM = case mode of
+           WithoutPadding -> Just msg
+           WithPadding -> paddingPKCS7 msg
+   msg' <- maybeToRight EmptyPayload msgM
+   mapBoth FromCryptonite (\c -> cbcEncrypt c initedIV msg') (initCipher key)
+
+-- | Decrypt using AES256 using CBC mode
+decrypt
+    :: CipherMode
+    -- ^ with or without padding
+    -> ByteString
+    -- ^ secret key, needs to be 32 bytes
+    -> ByteString
+    -- ^ iv, needs to be 16 bytes
+    -> ByteString
+    -- ^ payload, must be a multiple of a block size, ie., 16 bytes
+    -> Either CipherError ByteString
+decrypt mode key iv msg = do
+   when (mode == WithoutPadding && BS.length msg `mod` 16 /= 0) $
+       Left $ WrongPayloadSize
+   initedIV <- mapLeft FromCryptonite (initIV iv)
+   let unpadding p = case mode of
+           WithoutPadding -> Right p
+           WithPadding -> maybeToRight EmptyPayload (unpaddingPKCS7 p)
+   mapBoth FromCryptonite (\c -> cbcDecrypt c initedIV msg) (initCipher key) >>=
+       unpadding
+
+-- | Apply PKCS#7 padding to payload and end up with a multiple of a block size, i.e., 16 bytes,
+-- according to https://datatracker.ietf.org/doc/html/rfc5652#section-6.3 .
+-- The padding value is the number of padding bytes.
+-- If 1 byte of padding is required, the padding is "01".
+-- If 2 bytes of padding, it's "02 02".
+-- If no padding is required, an extra block of 0x10 bytes is added,
+-- i.e., sixteen copies of 16, which is the blocksize.
+-- This means that payload can only fit 15 bytes into a single block with padding.
+-- A 16 byte payload requires 2 blocks with padding.
+paddingPKCS7
+    :: ByteString
+    -> Maybe ByteString
+paddingPKCS7 payload =
+    let len = BS.length payload
+        remaining =
+            if len < 16 then
+                len
+            else
+                len `mod` 16
+        padding =
+            if remaining == 0 then
+                B8.replicate 16 (toEnum 16)
+            else
+                B8.replicate (16 - remaining) (toEnum (16 - remaining))
+    in if len == 0 then
+           Nothing
+       else
+           Just $ BS.append payload padding
+
+unpaddingPKCS7
+    :: ByteString
+    -> Maybe ByteString
+unpaddingPKCS7 payload =
+    let initLast = BS.unsnoc payload
+        cut (_, lastByte) =
+            if (fromEnum lastByte) >= BS.length payload then
+                payload
+            else
+                BS.dropEnd (fromEnum lastByte) payload
```

Comment:
Hi @paweljakubas 

Just curious -- why is the first branch of the `if` required? According to the [spec](https://www.ibm.com/docs/en/zos/2.4.0?topic=rules-pkcs-padding-method), couldn't we replace this with:

```suggestion
                let paddingLength = fromEnum lastByte in           
                BS.dropEnd paddingLength payload
```

Reply:
no no, everything passes. Meaning simplification you applied is correct!